elasticdl/pkg/commonnew/embedding_tabel.go

```diff
@@ -0,0 +1,70 @@
+package commonnew
+
+import (
+	"elasticdl.org/elasticdl/pkg/proto"
+	"github.com/tensorflow/tensorflow/tensorflow/go/core/framework/types_go_proto"
+)
+
+// EmbeddingTable struct
+type EmbeddingTable struct {
+	Dim              int64
+	Initializer      string
+	EmbeddingVectors map[int64]*Tensor
+	Dtype            types_go_proto.DataType
+}
+
+// NewIndexedSlices return proto.IndexedSlices
+func NewIndexedSlices(t *Tensor, indices []int64) *proto.IndexedSlices {
+	var i = proto.IndexedSlices{
+		ConcatTensors: t.SerializeTensor(),
+		Ids:           indices,
+	}
+	return &i
+}
+
+// NewEmbeddingTable creates an embedding table instance
+func NewEmbeddingTable(dim int64, initializer string, dtype types_go_proto.DataType) *EmbeddingTable {
+	var e = EmbeddingTable{
+		Dim:              dim,
+		Initializer:      initializer,
+		EmbeddingVectors: make(map[int64]*Tensor),
+		Dtype:            dtype,
+	}
+	return &e
+}
+
+// GetEmbeddingVector returns an REFERENCE of embedding vector giving an index
+func (e *EmbeddingTable) GetEmbeddingVector(index int64) *Tensor {
+	if value, ok := e.EmbeddingVectors[index]; ok {
+		return value
+	}
+	newVector := NewEmptyVector(e.Dim, e.Dtype)
+	e.EmbeddingVectors[index] = newVector
+	return newVector
+}
+
+// GetEmbeddingVectors returns COPYS of embedding vectors giving an array of indices
+func (e *EmbeddingTable) GetEmbeddingVectors(indices []int64) *Tensor {
+	dim := []int64{int64(len(indices)), e.Dim}
+	tensor := NewEmptyTensor(dim, e.Dtype)
+	for i, index := range indices {
+		if index == -1 {
+			continue
+		}
```

Comment:
What's the logic behind the code `if index == -1`?

Reply:
The logic here is, there will be a function `MergeIndexedSlice`, which will merge the same index column in-place, and the duplicate indices will be marked as -1, which indicates this should be ignored.

Reply:
The logic here is, there will be a function `MergeIndexedSlice`, which will merge the same index column in-place, and the duplicate indices will be marked as -1, which indicates this should be ignored.

Reply:
@lw396285v It seem a little hack here.

And we do not to merge gradients in PS side. I propose a #1749 , this PR add a logic in worker in worker to merge gradients of a same parameter already. So the gradients in PS are always unique.

So, please remove the `if index == -1` here int this PR. If we really need it,  let's find a more elegant way.

Reply:
fixed